src/RelayKnownVariableNames.ts

```diff
@@ -0,0 +1,66 @@
+import { ValidationRule } from "graphql"
+import { GraphQLError } from "./dependencies"
+import { getArgumentDefinitions } from "./argumentDefinitions"
+
+export function undefinedVarMessage(varName: string, opName?: string): string {
+  return opName
+    ? `Variable "$${varName}" is not defined by operation "${opName}".`
+    : `Variable "$${varName}" is not defined.`
+}
+
+export function undefinedVarMessageFragment(varName: string, fragmentName: string): string {
+  return `Variable "$${varName}" is not defined by fragment "${fragmentName}".`
+}
+
+export const RelayKnownVariableNames: ValidationRule = function RelayKnownVariableNames(context) {
+  let variableNameDefined = Object.create(null)
+  return {
+    OperationDefinition: {
+      enter() {
+        variableNameDefined = Object.create(null)
+      },
+      leave(operation) {
+        const usages = context.getVariableUsages(operation)
+
+        usages.forEach(({ node }) => {
+          const varName = node.name.value
+          if (variableNameDefined[varName] !== true) {
+            context.reportError(
+              new GraphQLError(undefinedVarMessage(varName, operation.name && operation.name.value), [node, operation])
+            )
+          }
+        })
+      },
+    },
+    VariableDefinition(node) {
+      variableNameDefined[node.variable.name.value] = true
+    },
+    FragmentDefinition: {
+      enter(fragmentDefinitionNode) {
+        variableNameDefined = Object.create(null)
+        const argumentDefinitions = getArgumentDefinitions(fragmentDefinitionNode)
+        if (argumentDefinitions != null) {
```

Comment:
Nit: I find that this just trips up people every now and then. Can it just be `if (argumentDefinitions) { … }`?

Reply:
Sure - I'm just used to our own coding conventions, but no problem :)

Reply:
I hear you, perhaps we should just have lint rules for such things and in the absence (I should) not nit-pick.